### PR TITLE
Use Tarn.js as a pool instead of generic-pool. fixes #1694 #1703

### LINKED ISSUE #1694: Connections are not returned to the pool if acquireConnectionTimeout is triggered
This test fails on 0.12.1

```js
const _ = require('lodash');
const Knex = require('knex');
const Promise = require('bluebird');

const knex = Knex({
  client: 'postgres',
  pool: {
    min: 0,
    max: 2
  },
  connection: {
    host: '127.0.0.1',
    database: 'test'
  },
  acquireConnectionTimeout: 100
});

// Start two transactions so that the pool is filled.
_.times(2, (idx) => {
  knex.transaction(() => {
    return Promise.delay(500);
  }).then(() => {
    console.log(`transaction ${idx} ended`);
  });
});

knex.raw('select 1').catch(() => {
  console.log('query failed as expected because of acquireConnectionTimeout');
});

setTimeout(() => {
  // By the time this is called, all the transactions should have been committed and
  // the connections returned to the pool, but the connection for which the acquireConnectionTimeout
  // was called, was never returned to the pool.
  console.log('connections in use:', knex.client.pool._inUseObjects.length, 'should be 0 right?');
}, 1500);
```

### PR DESCRIPTION
**Pull request by koskimas**
This pull request replaces the generic-pool with [Tarn.js](https://github.com/Vincit/tarn.js).

generic-pool doesn't have a built-in timeout for connection acquire nor does
it have a mechanism to cancel a pending acquire. Because of this the issue
#1694 is next to impossible to fix on the knex side.

generic-pool doesn't seem to be well maintained anymore. The last commit is
six monhts old at time of writing this and there are multiple pull requests
waiting, one of which is a fix for the acquire timeout needed here.

Tarn.js is a subset of generic-pool with added timeouts and a comprehensive
test suite. It is designed (by me) to be used with knex and I intend to make
that its only purpose. If you choose to merge this pull request. I'm committed 
to maintaining and bug fixing it to serve knex as well as possible. 

This pull request also fixes an issue #1703 where connections acquired for
transactions never time out, regardless of acquireConnectionTimeout

### CODE REVIEW

File: src/transaction.js [hunk 1]
```diff
@@ -142,6 +142,15 @@ export default class Transaction extends EventEmitter {
   acquireConnection(client, config, txid) {
     const configConnection = config && config.connection
     return Promise.try(() => configConnection || client.acquireConnection().completed)
+    .catch(TimeoutError, (error) => {
```
Comment (elhigu) [inline src/transaction.js]: This is nice addition, earlier I was able to fill my pool and stuck waiting new transaction forever...

File: src/client.js [hunk 2]
```diff
@@ -218,26 +219,18 @@ assign(Client.prototype, {
...
-    })
-    const abort = function(reason) {
-      if (request && !request.fulfilled) {
-        request.abort(reason)
+    } else {
+      return {
+        completed: this.pool.acquire().promise.then((connection) => {
+          debug('acquired connection from pool: %s', connection.__knexUid)
+          return connection
+        })
```
Comment (elhigu) [inline src/client.js]: Why abort is not needed anymore?

File: src/client.js [hunk 3]
```diff
@@ -252,17 +245,8 @@ assign(Client.prototype, {
...
-        return resolver()
-      }
-      this.pool.drain(() => {
-        this.pool.destroyAllNow(() => {
-          this.pool = undefined
-          resolver()
-        })
-      })
-    })
+    const promise = this.pool ? this.pool.destroy() : Promise.resolve()
```
Comment (elhigu) [inline src/client.js]: This seems to revert the change made 2 weeks ago, that destroying pool waits that all the queries are resolved before ending.

File: src/client.js (hunk 2 above)
Comment (koskimas) [inline src/client.js]: Tarn handles the acquire timeout. There is no need to handle it on the knex side anymore. There are tests for this.

File: src/client.js (hunk 3 above)
Comment (koskimas) [inline src/client.js]: Yeah, it was unclear to me what drain did. Currently Tarn murders all connections (both used and free) and calls the destroy function for them. 

I'll change destroy so that it waits for all connections to be released before destroying them.

File: src/client.js (hunk 2 above)
Comment (koskimas) [inline src/client.js]: This is fixed and tested in Tarn 0.1.5

File: src/client.js (hunk 3 above)
Comment (koskimas) [inline src/client.js]: This is now fixed in Tarn 0.1.5. I updated the package.json in the pull request.

**Comment by elhigu** [comment]
@tgriesser @rhys-vdw any any comments on this one? Pull request actually fixes some bugs + our own testsuites did work correctly after this (objection.js + one internal project backend with > 20k lines of typescript and which has 15k lines of tests).

**Comment by elhigu** [comment]
should fix this one too I suppose https://github.com/tgriesser/knex/issues/1709

**Comment by tgriesser** [comment]
Thanks for the PR... I'll try and get to reviewing this tomorrow, I'd personally rather make it so folks can specify their own pooling implementation rather than continue to switch it up. I'd also like to dig in a bit more about #1964 and see what all is happening there, I believe that abort stuff is legacy from pool2, so that probably needs to get fixed.

Not saying this won't get merged, just saying I'd like to investigate a bit further.

**Comment by koskimas** [comment]
I agree with you @tgriesser that switching back and forward with the pool isn't ideal and causes bugs. That is exactly what this pull request is all about. Finally having a pool that is written just for knex. I'd be happy to give you push rights to tarn.js.

Also allowing people to use their own pool is a good feature, but the default pool should serve most use cases.

The problem with #1694 was the legacy abort stuff, yes, but there is no mechanism in generic-pool to implement the same features the abort was solving. And pull requests to add the feature to generic-pool have been pending for six months. I would have just contributed to generic-pool if I thought that would solve this within next few months, but that doesn't seem to be the case.

I do realize that switching to a new library that has been around for a week is not the safest sounding idea 😄. This branch needs to be tested really well before making a release, if you decide to merge it. But I also think that resource pooling is not that difficult a problem. If the pool has good, meaningful tests it should be enough. There is no need for years of battle testing.

**Comment by koskimas** [comment]
And I'm not trying to rush you! Of course you should look into this, and thank you for doing that. Just putting all my thoughts out there 😄

**Comment by elhigu** [comment]
@tgriesser Any updates on this one? Current knex release is still leaking connections (which will fill the pool eventually) when ever aqcuireConnectionTimeout is triggered...

**Comment by tgriesser** [comment]
So 0.12.3 fixes the issue. Thanks for the PR but I think I'm going to just stick with generic-pool for now and instead focus on making it possible to supply the pool externally so folks can use their preferred pooling lib

**Comment by koskimas** [comment]
Can you give me a reason why this was not merged? Tarn has only the features knex uses, it is better tested than generic-pool and you I would have given the ownership for you in addition to committing to work on it myself. generic-pool doesn't seem to be maintained anymore.

**Comment by tgriesser** [comment]
A few reasons: 

- Was able to fix the reason for wanting to switch pools pretty easily in knex with a `setTimeout`
- Didn't have time to go through tarn, but was a bit confused by the recursion limit and the fact that tarn catches errors and logs them rather than forwarding them to the lib consumer. Also seems like there's a lot of promises for something which should generally be pretty simple.
- Not super concerned about the status of generic-pool maintenance since it generally works and is pretty well used across many libs
- Changing a dependency would require a minor version bump

Not ruling out revisiting this in the future if we have more pool issues, it just seemed simpler to fix it in knex land in this case. That said, I would like to make it simple to supply an external pool, will keep you updated on that.

**Comment by koskimas** [comment]
Fair enough. The most important thing is that the bug is now fixed, and knex is stable. Thank you for fixing it.

Recursion limit is there to stop the _asynchronous_ recursion in some cases when the pool is so jammed that it would retry to create a resource too many times. It is probably not needed, but I wanted to be extra paranoid.

Errors are logged only when there is no one to inform about the error. Whenever someone is waiting for resource, the resource promise is rejected. If you found a place where this is not the case, then it's a bug.

Promises are used in places where generic-pool would poll using `setTimeout`. Promises are more elegant than polling don't you think? Also isn't the point of using promises to make the async flow simpler? That is why I used promises instead of callbacks.

**Comment by elhigu** [comment]
IMO, this road where features, which should be responsibility of pool are implemented in knex side will come to an end at some point. Otherwise knex ends up implement all the meaningful functionality of pool. 

The biggest problem with generic pool is that lack of its maintenance forces knex to implement everything on knex side (e.g. maximum lease timeouts and maximum clients waiting to get lease).

Hope that this gets revisited at some point.

**Comment by koskimas** [comment]
@tgriesser Did you see my comment on the test case you added for the bug? It doesn't actually test anything. I used the `done` callback for a reason in the test. I'm aware that mocha accepts promises.

**Comment by fl0w** [comment]
Since this didn't get merged, @koskimas would you be against me publishing a monkey-patch-package, e.g. `knex-tarn`?

**Comment by koskimas** [comment]
Currently tarn doesn't bring any new functionality to knex. Since @tgriesser fixed the bugs only reason to merge tarn is future development. I don't think we should create such a fork. Thanks for the offer though.

**Comment by tgriesser** [comment]
@koskimas yep I updated the test. Also if you want to PR adding tarn against the refactor branch, I'd be up for exploring it a bit further... Specifically would like to see if tarn helps with the `unhandledRejection` I can't seem to figure out the source of in that branch.
